src/QueryBuilder.ts

```diff
@@ -6,219 +8,223 @@ import {
 
-    private _prepaymentValidate(errors: string[]): void {
-        this._doValidate(errors);
-    }
+            const respHeader = this._mapResponseHeader(resp);
+            throwIfExceptional(respHeader.getNodetransactionprecheckcode());
 
-    /**
-     * Request the cost of this query in HBAR from the node.
-     *
-     * You can then attach a payment for this value with `.setPaymentAmount()`.
-     */
-    public async getCost(client: BaseClient): Promise<Hbar> {
-        runValidation(this, (errors) => this._prepaymentValidate(errors));
-
-        // create a duplicate of the query with `COST_ANSWER` instead of the original response type
-        // we also must have a signed payment of 0 hbar which is not actually processed
-        const responseType = this._header.getResponsetype();
-        this._header.setResponsetype(ResponseType.COST_ANSWER);
+            return Hbar.fromTinybar(respHeader.getCost());
+        } finally {
+            // Reset the response type and payment transaction
+            // on the query header
+            queryHeader.setResponsetype(currentResponseType);
+            queryHeader.setPayment(currentPayment);
+        }
+    }
 
-        const payment = this._header.getPayment();
-        await this._generatePayment(0, client);
+    public execute(client: BaseClient): Promise<T> {
+        let respStatus: ResponseCode | null = null;
 
-        const query = this._inner.clone();
+        return timeoutPromise(this._getDefaultExecuteTimeout(), (async() => {
+            let node: Node;
 
-        this._header.setResponsetype(responseType);
-        this._header.setPayment(payment);
+            if (this._isPaymentRequired()) {
+                if (this._getHeader().hasPayment()) {
+                    const paymentTxBodyBytes = this._getHeader().getPayment()!.getBodybytes_asU8();
+                    const paymentTxBody = TransactionBody.deserializeBinary(paymentTxBodyBytes);
 
-        const node = this._getNode(client);
+                    const nodeId = AccountId._fromProto(paymentTxBody.getNodeaccountid()!);
 
-        const response = await client._unaryCall(node.url, query, this._method);
+                    node = client._getNode(nodeId);
+                } else if (this._paymentAmount != null) {
+                    node = client._randomNode();
 
-        const responseHeader = getResponseHeader(response);
-        throwIfExceptional(responseHeader.getNodetransactionprecheckcode());
+                    await this._generatePaymentTransaction(client, node, this._paymentAmount);
+                } else if (this._maxPaymentAmount != null || client.maxQueryPayment != null) {
+                    node = client._randomNode();
 
-        return Hbar.fromTinybar(responseHeader.getCost());
-    }
+                    // TODO[2020-01-01]: Use `??` when we can
+                    const maxPaymentAmount: Hbar = this._maxPaymentAmount ||
+                        client.maxQueryPayment!;
```

Comment:
I think we made `client.maxQueryPayment` default to 1 hbar and then the user can set it to 0 to disable payments.

Reply:
It does actually now that I double check the Client class.